component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/voiceotp/VoiceOTPAuthenticator.java

```diff
@@ -0,0 +1,2037 @@
+
+            log.debug("Default authentication endpoint context is used");
+
+        }
+        return loginPage;
+    }
+
+    /**
+     * Get the errorPage from authentication.xml file or use the error page from constant file.
+     *
+     * @param context the AuthenticationContext
+     * @return the errorPage
+     * @throws AuthenticationFailedException
+     */
+    private String getErrorPage(AuthenticationContext context) throws AuthenticationFailedException {
+
+        String errorPage = VoiceOTPUtils.getErrorPageFromXMLFile(context);
+        if (StringUtils.isEmpty(errorPage)) {
+            errorPage = ConfigurationFacade.getInstance().getAuthenticationEndpointURL()
+                    .replace(VoiceOTPConstants.LOGIN_PAGE, VoiceOTPConstants.ERROR_PAGE);
+
+            log.debug("Default authentication endpoint context is used");
+
+        }
+        return errorPage;
+    }
+
+    /**
+     * To get the redirection URL.
+     *
+     * @param baseURI     the base path
+     * @param queryParams the queryParams
+     * @return url
+     */
+    private String getURL(String baseURI, String queryParams) {
+
+        String url;
+        if (StringUtils.isNotEmpty(queryParams)) {
+            url = baseURI + "?" + queryParams + "&" + VoiceOTPConstants.NAME_OF_AUTHENTICATORS + getName();
+        } else {
+            url = baseURI + "?" + VoiceOTPConstants.NAME_OF_AUTHENTICATORS + getName();
+        }
+        return url;
+    }
+
+    /**
+     * Redirect to an error page.
+     *
+     * @param response    the HttpServletResponse
+     * @param queryParams the queryParams
+     * @throws AuthenticationFailedException
+     */
+    private void redirectToErrorPage(HttpServletResponse response, AuthenticationContext context, String queryParams,
+                                     String retryParam)
+            throws AuthenticationFailedException {
+        // that Enable the Voice OTP in user's Profile. Cannot proceed further without Voice OTP authentication.
+        try {
+            String errorPage = getErrorPage(context);
+            String url = getURL(errorPage, queryParams);
+            response.sendRedirect(url + retryParam);
+        } catch (IOException e) {
+            throw new AuthenticationFailedException("Exception occurred while redirecting to errorPage. ", e);
+        }
+    }
+
+    /**
+     * In VoiceOTP optional case proceed with first step only.It can be basic or federated.
+     *
+     * @param authenticatedUser the name of authenticatedUser
+     * @param context           the AuthenticationContext
+     */
+    private void processFirstStepOnly(AuthenticatedUser authenticatedUser, AuthenticationContext context) {
+
+
+        log.debug("Processing First step only. Skipping VoiceOTP");
+
+        //the authentication flow happens with basic authentication.
+        StepConfig stepConfig = context.getSequenceConfig().getStepMap().get(context.getCurrentStep() - 1);
```

Comment:
`context.getSequenceConfig().getStepMap()` could this be null?

Reply:
This becomes null in request path based authentication only.